pkg/semtech-loramac/contrib/semtech_loramac_radio.c

```diff
@@ -106,92 +138,108 @@ void SX127XSetTxConfig(RadioModems_t modem, int8_t power, uint32_t fdev,
 {
-    (void) fdev;
-    (void) fixLen;
-    sx127x_set_modem(&sx127x, modem);
-    sx127x_set_freq_hop(&sx127x, freqHopOn);
-    sx127x_set_bandwidth(&sx127x, bandwidth);
-    sx127x_set_coding_rate(&sx127x, coding_rate);
-    sx127x_set_spreading_factor(&sx127x, spreading_factor);
-    sx127x_set_crc(&sx127x, crcOn);
-    sx127x_set_freq_hop(&sx127x, freqHopOn);
-    sx127x_set_hop_period(&sx127x, hopPeriod);
-    sx127x_set_fixed_header_len_mode(&sx127x, false);
-    sx127x_set_iq_invert(&sx127x, iqInverted);
-    sx127x_set_payload_length(&sx127x, 0);
-    sx127x_set_tx_power(&sx127x, power);
-    sx127x_set_preamble_length(&sx127x, preambleLen);
-    sx127x_set_rx_single(&sx127x, false);
-    sx127x_set_tx_timeout(&sx127x, timeout); /* base unit ms, LoRaMAC ms */
+    (void)fdev;
+    (void)fixLen;
+    (void)modem;
+    uint16_t modem_val = NETDEV_TYPE_LORA;
+    loramac_netdev_ptr->driver->set(loramac_netdev_ptr, NETOPT_DEVICE_TYPE, &modem_val, sizeof(uint16_t));
+    loramac_netdev_ptr->driver->set(loramac_netdev_ptr, NETOPT_BANDWIDTH, &bandwidth, sizeof(uint8_t));
+    loramac_netdev_ptr->driver->set(loramac_netdev_ptr, NETOPT_CODING_RATE, &coding_rate, sizeof(uint8_t));
+    loramac_netdev_ptr->driver->set(loramac_netdev_ptr, NETOPT_SPREADING_FACTOR, &spreading_factor, sizeof(uint8_t));
+    netopt_enable_t enable_crc = (crcOn) ? NETOPT_ENABLE : NETOPT_DISABLE;
+    loramac_netdev_ptr->driver->set(loramac_netdev_ptr, NETOPT_INTEGRITY_CHECK, &enable_crc, sizeof(netopt_enable_t));
+    netopt_enable_t enable_freq_hop = (freqHopOn) ? NETOPT_ENABLE : NETOPT_DISABLE;
+    loramac_netdev_ptr->driver->set(loramac_netdev_ptr, NETOPT_CHANNEL_HOP, &enable_freq_hop, sizeof(netopt_enable_t));
+    loramac_netdev_ptr->driver->set(loramac_netdev_ptr, NETOPT_CHANNEL_HOP_PERIOD, &hopPeriod, sizeof(uint8_t));
+    netopt_enable_t fixed_header = NETOPT_DISABLE;
+    loramac_netdev_ptr->driver->set(loramac_netdev_ptr, NETOPT_FIXED_HEADER, &fixed_header, sizeof(netopt_enable_t));
+    netopt_enable_t iq_inverted = (iqInverted) ? NETOPT_ENABLE : NETOPT_DISABLE;
+    loramac_netdev_ptr->driver->set(loramac_netdev_ptr, NETOPT_IQ_INVERT, &iq_inverted, sizeof(netopt_enable_t));
+    int16_t power_val = power;
+    loramac_netdev_ptr->driver->set(loramac_netdev_ptr, NETOPT_TX_POWER, &power_val, sizeof(int16_t));
+    loramac_netdev_ptr->driver->set(loramac_netdev_ptr, NETOPT_PREAMBLE_LENGTH, &preambleLen, sizeof(uint16_t));
+    loramac_netdev_ptr->driver->set(loramac_netdev_ptr, NETOPT_TX_TIMEOUT, &timeout, sizeof(uint32_t)); /* base unit ms, LoRaMAC ms */
 }
 
 uint32_t SX127XTimeOnAir(RadioModems_t modem, uint8_t pktLen)
 {
     (void) modem;
-    return sx127x_get_time_on_air(&sx127x, pktLen);
+    uint8_t cr;
+    loramac_netdev_ptr->driver->get(loramac_netdev_ptr, NETOPT_CODING_RATE, &cr, sizeof(uint8_t));
+    MibRequestConfirm_t mibReq;
+    mibReq.Type = MIB_CHANNELS_DEFAULT_DATARATE;
```

Comment:
```suggestion
    mibReq.Type = MIB_CHANNELS_DATARATE;
```

Otherwise you will always get the same "default" channel

Reply:
This is the default datarate, not channel, but I guess this is ok to use `MIB_CHANNELS_DATARATE` at this level since the datarate might be changed internally by the MAC.

Reply:
Sorry, I meant datarate

Reply:
Anyway, the code seems weird (`MIB_CHANNELS_DATARATE` here and `mibReq.Param.ChannelsDatarate` just below), si I think your comment makes sense. I'm testing with `MIB_CHANNELS_DATARATE`, just to be sure.

Reply:
<details>

```
$ DEVEUI_ABP=00EA17067F596D18 APPSKEY=E38E70FD74D768BE36EA753D96D5030C NWKSKEY=C89D06AB37D58373FA50B69B7D288B1C DEVADDR=26011C28 RX2_DR=3 DEVEUI_OTA=00AB22849FF4885C APPEUI=70B3D57ED000B02F APPKEY=45C8E67A4FD8871D6943CEE3A305177C IOTLAB_NODE=auto-ssh make -C tests/pkg_semtech-loramac/ flash test-with-config --no-print-directory 
Building application "tests_pkg_semtech-loramac" for "b-l072z-lrwan1" with MCU "stm32".

"make" -C /work/riot/RIOT/pkg/semtech-loramac
"make" -C /work/riot/RIOT/build/pkg/semtech-loramac/src/boards/mcu -f /work/riot/RIOT/pkg/semtech-loramac/Makefile.loramac_arch
"make" -C /work/riot/RIOT/build/pkg/semtech-loramac/src/system/crypto -f /work/riot/RIOT/pkg/semtech-loramac/Makefile.loramac_crypto
"make" -C /work/riot/RIOT/build/pkg/semtech-loramac/src/mac -f /work/riot/RIOT/pkg/semtech-loramac/Makefile.loramac_mac
"make" -C /work/riot/RIOT/build/pkg/semtech-loramac/src/mac/region -f /work/riot/RIOT/pkg/semtech-loramac/Makefile.loramac_region
"make" -C /work/riot/RIOT/boards/b-l072z-lrwan1
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/cpu/stm32
"make" -C /work/riot/RIOT/cpu/cortexm_common
"make" -C /work/riot/RIOT/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT/cpu/stm32/periph
"make" -C /work/riot/RIOT/cpu/stm32/stmclk
"make" -C /work/riot/RIOT/cpu/stm32/vectors
"make" -C /work/riot/RIOT/drivers
"make" -C /work/riot/RIOT/drivers/periph_common
"make" -C /work/riot/RIOT/drivers/sx127x
"make" -C /work/riot/RIOT/pkg/semtech-loramac/contrib
"make" -C /work/riot/RIOT/sys
"make" -C /work/riot/RIOT/sys/auto_init
"make" -C /work/riot/RIOT/sys/auto_init/loramac
"make" -C /work/riot/RIOT/sys/fmt
"make" -C /work/riot/RIOT/sys/frac
"make" -C /work/riot/RIOT/sys/iolist
"make" -C /work/riot/RIOT/sys/isrpipe
"make" -C /work/riot/RIOT/sys/luid
"make" -C /work/riot/RIOT/sys/malloc_thread_safe
"make" -C /work/riot/RIOT/sys/newlib_syscalls_default
"make" -C /work/riot/RIOT/sys/pm_layered
"make" -C /work/riot/RIOT/sys/random
"make" -C /work/riot/RIOT/sys/random/tinymt32
"make" -C /work/riot/RIOT/sys/shell
"make" -C /work/riot/RIOT/sys/shell/commands
"make" -C /work/riot/RIOT/sys/stdio_uart
"make" -C /work/riot/RIOT/sys/test_utils/interactive_sync
"make" -C /work/riot/RIOT/sys/tsrb
"make" -C /work/riot/RIOT/sys/ztimer
   text	   data	    bss	    dec	    hex	filename
  59212	    292	   7360	  66864	  10530	/work/riot/RIOT/tests/pkg_semtech-loramac/bin/b-l072z-lrwan1/tests_pkg_semtech-loramac.elf
For full programmer output add PROGRAMMER_QUIET=0 or QUIET=0 to the make command line.
 ✓ Flashing done! (programmer: 'iotlab' - duration: 12.04s)


ssh -t abadie@saclay.iot-lab.info 'socat - tcp:st-lrwan1-1.saclay.iot-lab.info:20000' 
main(): This is RIOT! (Version: 2021.04-devel-1043-g90b74-semtech-loramac_netdev)
All up, running the shell now
> loramac erase

> 
> loramac erase
> reboot
reboot
main(): This is RIOT! (Version: 2021.04-devel-1043-g90b74-semtech-loramac_netdev)
All up, running the shell now
> loramac set deveui 00AB22849FF4885C
loramac set deveui 00AB22849FF4885C
> loramac set appeui 70B3D57ED000B02F
loramac set appeui 70B3D57ED000B02F
> loramac set appkey 45C8E67A4FD8871D6943CEE3A305177C
loramac set appkey 45C8E67A4FD8871D6943CEE3A305177C
> loramac set dr 0
loramac set dr 0
> loramac join otaa
loramac join otaa
Join procedure succeeded!
> loramac tx "This is RIOT" cnf 123
loramac tx "This is RIOT" cnf 123
Received ACK from network
Message sent with success
> loramac tx "This is RIOT" uncnf 42
loramac tx "This is RIOT" uncnf 42
Message sent with success
> reboot
reboot
main(): This is RIOT! (Version: 2021.04-devel-1043-g90b74-semtech-loramac_netdev)
All up, running the shell now
> loramac set deveui 00AB22849FF4885C
loramac set deveui 00AB22849FF4885C
> loramac set appeui 70B3D57ED000B02F
loramac set appeui 70B3D57ED000B02F
> loramac set appkey 45C8E67A4FD8871D6943CEE3A305177C
loramac set appkey 45C8E67A4FD8871D6943CEE3A305177C
> loramac set dr 3
loramac set dr 3
> loramac join otaa
loramac join otaa
Join procedure succeeded!
> loramac tx "This is RIOT" cnf 123
loramac tx "This is RIOT" cnf 123
Received ACK from network
Message sent with success
> loramac tx "This is RIOT" uncnf 42
loramac tx "This is RIOT" uncnf 42
Message sent with success
> reboot
reboot
main(): This is RIOT! (Version: 2021.04-devel-1043-g90b74-semtech-loramac_netdev)
All up, running the shell now
> loramac set deveui 00AB22849FF4885C
loramac set deveui 00AB22849FF4885C
> loramac set appeui 70B3D57ED000B02F
loramac set appeui 70B3D57ED000B02F
> loramac set appkey 45C8E67A4FD8871D6943CEE3A305177C
loramac set appkey 45C8E67A4FD8871D6943CEE3A305177C
> loramac set dr 5
loramac set dr 5
> loramac join otaa
loramac join otaa
Join procedure succeeded!
> loramac tx "This is RIOT" cnf 123
loramac tx "This is RIOT" cnf 123
Received ACK from network
Message sent with success
> loramac tx "This is RIOT" uncnf 42
loramac tx "This is RIOT" uncnf 42
Message sent with success
> reboot
reboot
main(): This is RIOT! (Version: 2021.04-devel-1043-g90b74-semtech-loramac_netdev)
All up, running the shell now
> loramac erase
loramac erase
> reboot
reboot
main(): This is RIOT! (Version: 2021.04-devel-1043-g90b74-semtech-loramac_netdev)
All up, running the shell now
> loramac set deveui 00EA17067F596D18
loramac set deveui 00EA17067F596D18
> loramac set appeui 70B3D57ED000B02F
loramac set appeui 70B3D57ED000B02F
> loramac set devaddr 26011C28
loramac set devaddr 26011C28
> loramac set nwkskey C89D06AB37D58373FA50B69B7D288B1C
loramac set nwkskey C89D06AB37D58373FA50B69B7D288B1C
> loramac set appskey E38E70FD74D768BE36EA753D96D5030C
loramac set appskey E38E70FD74D768BE36EA753D96D5030C
> loramac set rx2_dr 3
loramac set rx2_dr 3
> loramac set dr 0
loramac set dr 0
> loramac join abp
loramac join abp
Join procedure succeeded!
> loramac tx "This is RIOT" cnf 123
loramac tx "This is RIOT" cnf 123
Received ACK from network
Message sent with success
> loramac tx "This is RIOT" uncnf 42
loramac tx "This is RIOT" uncnf 42
Message sent with success
>loramac get ul_cnt
 loramac get ul_cnt
Uplink Counter: 2
> reboot
reboot
main(): This is RIOT! (Version: 2021.04-devel-1043-g90b74-semtech-loramac_netdev)
All up, running the shell now
> loramac set ul_cnt 2
loramac set ul_cnt 2
> loramac set deveui 00EA17067F596D18
loramac set deveui 00EA17067F596D18
> loramac set appeui 70B3D57ED000B02F
loramac set appeui 70B3D57ED000B02F
> loramac set devaddr 26011C28
loramac set devaddr 26011C28
>loramac set nwkskey C89D06AB37D58373FA50B69B7D288B1C
 loramac set nwkskey C89D06AB37D58373FA50B69B7D288B1C
> loramac set appskey E38E70FD74D768BE36EA753D96D5030C
loramac set appskey E38E70FD74D768BE36EA753D96D5030C
> loramac set rx2_dr 3
loramac set rx2_dr 3
> loramac set dr 3
loramac set dr 3
> loramac join abp
loramac join abp
Join procedure succeeded!
> loramac tx "This is RIOT" cnf 123
loramac tx "This is RIOT" cnf 123
Received ACK from network
Message sent with success
> loramac tx "This is RIOT" uncnf 42
loramac tx "This is RIOT" uncnf 42
Message sent with success
> loramac get ul_cnt
loramac get ul_cnt
Uplink Counter: 4
> reboot
reboot
main(): This is RIOT! (Version: 2021.04-devel-1043-g90b74-semtech-loramac_netdev)
All up, running the shell now
> loramac set ul_cnt 4
loramac set ul_cnt 4
> loramac set deveui 00EA17067F596D18
loramac set deveui 00EA17067F596D18
> loramac set appeui 70B3D57ED000B02F
loramac set appeui 70B3D57ED000B02F
> loramac set devaddr 26011C28
loramac set devaddr 26011C28
> loramac set nwkskey C89D06AB37D58373FA50B69B7D288B1C
loramac set nwkskey C89D06AB37D58373FA50B69B7D288B1C
> loramac set appskey E38E70FD74D768BE36EA753D96D5030C
loramac set appskey E38E70FD74D768BE36EA753D96D5030C
> loramac set rx2_dr 3
loramac set rx2_dr 3
> loramac set dr 5
loramac set dr 5
> loramac join abp
loramac join abp
Join procedure succeeded!
> loramac tx "This is RIOT" cnf 123
loramac tx "This is RIOT" cnf 123
Received ACK from network
Message sent with success
> loramac tx "This is RIOT" uncnf 42
loramac tx "This is RIOT" uncnf 42
Message sent with success
> loramac get ul_cnt
loramac get ul_cnt
Uplink Counter: 6reboot

> reboot
main(): This is RIOT! (Version: 2021.04-devel-1043-g90b74-semtech-loramac_netdev)
All up, running the shell now
> loramac set ul_cnt 6
loramac set ul_cnt 6
> loramac help
loramac help
Usage: loramac <get|set|join|tx|link_check|save|erase>
> loramac erase
loramac erase
> reboot
reboot
main(): This is RIOT! (Version: 2021.04-devel-1043-g90b74-semtech-loramac_netdev)
All up, running the shell now
> loramac get deveui
loramac get deveui
DEVEUI: 0000000000000000
> loramac get appeui
loramac get appeui
APPEUI: 0000000000000000
> loramac get appkey
loramac get appkey
APPKEY: 00000000000000000000000000000000
> loramac get devaddr
loramac get devaddr
DEVADDR: 00000000
> loramac get nwkskey
loramac get nwkskey
NWKSKEY: 00000000000000000000000000000000
> loramac get appskey
loramac get appskey
APPSKEY: 00000000000000000000000000000000
>loramac set deveui 00AB22849FF4885C
 loramac set deveui 00AB22849FF4885C
> loramac set appeui 70B3D57ED000B02F
loramac set appeui 70B3D57ED000B02F
> loramac set appkey 45C8E67A4FD8871D6943CEE3A305177C
loramac set appkey 45C8E67A4FD8871D6943CEE3A305177C
> loramac save
loramac save
> reboot
reboot
main(): This is RIOT! (Version: 2021.04-devel-1043-g90b74-semtech-loramac_netdev)
All up, running the shell now
> loramac get deveui
loramac get deveui
DEVEUI: 00AB22849FF4885C
> loramac get appeui
loramac get appeui
APPEUI: 70B3D57ED000B02F
> loramac get appkey
loramac get appkey
APPKEY: 45C8E67A4FD8871D6943CEE3A305177C
> loramac erase
loramac erase
> reboot
reboot
main(): This is RIOT! (Version: 2021.04-devel-1043-g90b74-semtech-loramac_netdev)
All up, running the shell now
> loramac set deveui 00EA17067F596D18
loramac set deveui 00EA17067F596D18
> loramac set appeui 70B3D57ED000B02F
loramac set appeui 70B3D57ED000B02F
> loramac set devaddr 26011C28
loramac set devaddr 26011C28
> loramac set nwkskey C89D06AB37D58373FA50B69B7D288B1C
loramac set nwkskey C89D06AB37D58373FA50B69B7D288B1C
> loramac set appskey E38E70FD74D768BE36EA753D96D5030C
loramac set appskey E38E70FD74D768BE36EA753D96D5030C
> loramac set rx2_dr 3
loramac set rx2_dr 3
> loramac save
loramac save
> reboot
reboot
main(): This is RIOT! (Version: 2021.04-devel-1043-g90b74-semtech-loramac_netdev)
All up, running the shell now
> loramac get devaddr
loramac get devaddr
DEVADDR: 26011C28
> loramac get nwkskey
loramac get nwkskey
NWKSKEY: C89D06AB37D58373FA50B69B7D288B1C
> loramac get appskey
loramac get appskey
APPSKEY: E38E70FD74D768BE36EA753D96D5030C
> TEST PASSED

```

</details>